### PR TITLE
Improve BERT tokenization for accented characters and non-latin scripts

### PR DESCRIPTION
**Pull request by iamlemec**
I believe this addresses most of the issues involved in #5496. Adds in a (big) table to `unicode.h` with NFD codepoint mappings for accented characters and uses this in tokenization. Also addresses some other issues regarding conversion to lowercase in greek and cyrillic as well as puncuation padding.

### CODE REVIEW

**Comment by cebtenzzre** [comment]
Much better, though there are still some differences:
```diff
--- good_tokens.txt	2024-02-26 16:40:56.425215339 -0500
+++ lcpp_tokens.txt	2024-02-26 16:40:56.757218189 -0500
@@ -200554,7 +200554,6 @@
 1337: ष
 29870: ##ल
 29869: ##र
-29879: ##ो
 29863: ##न
 1317: ग
 1000: "
@@ -200633,7 +200632,11 @@
 29836: ##و
 29817: ##ت
 25573: ##ا
-100: [UNK]
+1282: س
+23673: ##ل
+29836: ##و
+15394: ##د
+29836: ##و
 23856: kota
 16183: sal
 6784: ##ud
@@ -254648,7 +254651,6 @@
 3747: influence
 1997: of
 2332: king
-1097: æ
 10760: ##the
 20850: ##lb
 19058: ##ald
```

1. `षेलुरोङ्` from the Wikipedia page on Manila becomes `षेलुरोन ् ग ्` via wiki.test.raw mojibake. Not sure that it matters.
2. `كوتا سلودوڠ` from the same page is ignored by the official implementation for some reason (probably a bug?)
3. `Æthelbald` from the page on Tatwine, `Æ` is dropped instead of being correctly converted to lowercase.

**Comment by cebtenzzre** [comment]
Also, can we format nfd_map better in the source code? At a quick glance, it doesn't even appear to be sorted.

File: llama.cpp [hunk 1]
```diff
@@ -8897,37 +8897,45 @@ struct llm_tokenizer_wpm {
...
 
     std::vector<std::string> preprocess(const std::string & text) {
-        std::string ori_str = normalize(text);
-        uint64_t ori_size = ori_str.size();
+        // normalalization form D
+        std::vector<uint32_t> codepoints = codepoints_from_utf8(text);
+        std::vector<uint32_t> nfd_codepoints;
+        for (uint32_t code : codepoints) {
+            if (nfd_map.find(code) != nfd_map.end()) {
+                for (uint32_t c : nfd_map[code]) {
```
[the remark this answers is not in the document]
Comment (iamlemec) [reply llama.cpp]: Yup, did `it->second`.

**Comment by iamlemec** [comment]
Sorted the NFD map and capped at 200 chars width. As for the tokenization variances:

1. Looks like ``ो`` is an accent character and we properly drop it.
2. Yeah, I think we're closer to correct here.
3. Right now I'm just hard coding standard Latin, Greek, and Cyrillic letters for lowercasing. There doesn't seem to be any systematic list of lowercase-able letters I can find. Open to ideas here.

Other than that, seeing some other variances, such as `Đ` doesn't have an NFD entry and various Devangari cases, but overall looking good.

**Review by ggerganov — APPROVED** [review]

**Comment by cebtenzzre** [comment]
> 3\. There doesn't seem to be any systematic list of lowercase-able letters I can find. Open to ideas here.

The Unicode standard has a lot to say about this: https://unicode.org/faq/casemap_charprop.html#casemap

There must be an MIT-licensed case mapping implementation somewhere that we could use if we wanted to be more correct.

**Review by cebtenzzre — APPROVED** [review]

**Comment by cebtenzzre** [comment]
This *is* C++ source, why not use the standard library for casing operations? https://en.cppreference.com/w/cpp/locale/tolower

**Comment by iamlemec** [comment]
Excellent idea. Ok, so I'm going to use:
```c++
std::tolower((wchar_t)code, std::locale("en_US.UTF-8"))
```
Works on Unix as `wchar_t` is 32 bits, but on Windows `wchar_t` is only 16 bits, so we just don't touch anything above `0xFFFF`. There are a very small number of capital letters above `0xFFFF`, but they don't appear to be from any scripts currently in wide use. Also checked the BGE and Nomic codepoints and none are in that range.

**Comment by cebtenzzre** [comment]
> Works on Unix as `wchar_t` is 32 bits, but on Windows `wchar_t` is only 16 bits, so we just don't touch anything above `0xFFFF`.

C++11 has char32_t which I think does what you need.

File: llama.cpp [hunk 2]
```diff
@@ -68,6 +68,7 @@
 #include <cstdio>
 #include <cstring>
 #include <ctime>
+#include <cwctype>
```
Comment (cebtenzzre) [inline llama.cpp:71]: Should also `#include <locale>` for std::locale and the locale-aware version of tolower.

**Comment by iamlemec** [comment]
For some reason if I cast it to anything other than `wchar_t`, it will terminate on a `std::bad_cast` (on Linux). I don't have access to a Windows machine, so I can't really test different options there. Though the CI says the current version at least compiles.

File: llama.cpp (hunk 2 above)
Comment (iamlemec) [reply llama.cpp:71]: Yup.

**Comment by cebtenzzre** [comment]
> For some reason if I cast it to anything other than `wchar_t`, it will terminate on a `std::bad_cast` (on Linux).

Apparently even as of C++20, char32_t is only implemented for the std::codecvt facet, not the std::ctype facet. use_facet will throw std::bad_cast if the locale does not provide the given facet, so that's what you are seeing.

File: llama.cpp [hunk 3]
```diff
@@ -8945,47 +8956,22 @@ struct llm_tokenizer_wpm {
...
+    uint32_t to_lower(uint32_t code) {
+#if defined(_WIN32)
+        if (code > 0xFFFF) {
+            return code;
         }
-        return text2;
+        return std::tolower(static_cast<wchar_t>(code), std::locale("en_US.UTF-8"));
+#else
+        return std::tolower((wchar_t)code, std::locale("en_US.UTF-8"));
+#endif
```
Comment (cebtenzzre) [inline llama.cpp:8967]:
```suggestion
#endif
        return std::tolower(wchar_t(code), std::locale("en_US.UTF-8"));
```

**Review by cebtenzzre — APPROVED** [review]

**Comment by ggerganov** [comment]
@iamlemec 

This PR increased the compile time for `llama.cpp` source file from 5s to 30s on M2 Ultra:

```bash
time c++ -std=c++11 -fPIC -O3 -pthread  -I. -Icommon -D_XOPEN_SOURCE=600 -D_DARWIN_C_SOURCE -DNDEBUG -c llama.cpp -o llama.o
```

``` bash
# before
real	0m5.506s
user	0m5.352s
sys	    0m0.101s

# after
real	0m29.161s
user	0m27.767s
sys	    0m1.029s
```

We should fix this ASAP. The problem stems from the large NFD map

**Comment by iamlemec** [comment]
Yup, seeing roughly 8s -> 20s on Linux. I suppose it's not happy about the map-vector setup. Looking into alternative data structures now.

**Comment by iamlemec** [comment]
Kinda tricky, but I have it working and back to normal compile times when `nfd_map` is
```c++
typedef std::pair<uint32_t, uint32_t> CodePos;
struct CodePosCompare {
    bool operator()(const CodePos & lhs, const CodePos & rhs) const {
        return lhs.first < rhs.first;
    }
};
static const std::multimap<CodePos, uint32_t, CodePosCompare> nfd_map = {
    {{0xC0, 0}, 0x41}, {{0xC0, 1}, 0x300}, {{0xC1, 0}, 0x41}, {{0xC1, 1}, 0x301}, ...
};
```
The docs say that the insertion order is respected when items are equivalent, so `equal_range` will give the correct results.

File: llama.cpp [hunk 4]
```diff
@@ -8945,47 +8956,20 @@ struct llm_tokenizer_wpm {
...
-                text2[i] = c - 'A' + 'a';
-            }
+    uint32_t to_lower(uint32_t code) {
+#if defined(_WIN32)
+        if (code > 0xFFFF) {
+            return code;
         }
-        return text2;
+#endif
+        return std::tolower(wchar_t(code), std::locale("en_US.UTF-8"));
```
Comment (cebtenzzre) [inline llama.cpp:8965]: @iamlemec This depends on the system having the `en_US.UTF-8` locale available, which is not always true in practice: https://github.com/nomic-ai/gpt4all/issues/2160

It seems like `.UTF-8` at least works on Windows (it's used by llama-bench), but on Unix I suppose you can only really assume that the plain C locale is available...

Comment (cebtenzzre) [reply llama.cpp:8965]: Also, it's not safe to assume that lowercasing one letter at a time is the same as lowercasing the whole string - see the exception for capital sigma at the end of a word https://github.com/yhirose/cpp-unicodelib/blob/797b1f0f1592ce13afabf3576f51ef26db5e884d/UCD/SpecialCasing.txt#L211

**Pull request merged**